src/mlpack/methods/ann/rbm/binary_rbm_policy.hpp

```diff
@@ -0,0 +1,162 @@
+   */
+  void HiddenMean(InputDataType&& input, OutputDataType&& output);
+
+  /**
+   * SampleVisible function samples the visible 
+   * layer using bernoulli function.
+   *
+   * @param input hidden neurons
+   * @param output sampled visible neurons
+   */
+  void SampleVisible(InputDataType&& input, OutputDataType&& output);
+
+  /**
+   * SampleHidden function samples the hidden 
+   * layer using bernoulli function.
+   *
+   * @param input visible neurons
+   * @param output sampled hidden neurons
+   */
+  void SampleHidden(InputDataType&& input, OutputDataType&& output);
+
+  template<typename Archive>
+  void Serialize(Archive& ar, const unsigned int /* version */);
+
+  //! Return the initial point for the optimization.
+  const InputDataType& Parameters() const { return parameter; }
+  //! Modify the initial point for the optimization.
+  InputDataType& Parameters() { return parameter; }
+  //! Return the weights of the network
+  const OutputDataType& Weight() const { return weight; }
+  //! Modify the weight of the network
+  OutputDataType& Weight() { return weight; }
+
+  //! Return the visible bias of the network
+  const InputDataType& VisibleBias() const { return visibleBias; }
+  //! Modify the visible bias of the network
+  InputDataType& VisibleBias() { return visibleBias; }
+
+  //! Return the hidden bias of the network
+  const InputDataType& HiddenBias() const { return hiddenBias; }
+  //! Modify the  hidden bias of the network
+  InputDataType& HiddenBias() { return hiddenBias; }
+
+  //! Get the visible size
+  size_t const& VisibleSize() const { return visibleSize; }
+  //! Get the hidden size
+  size_t const& HiddenSize() const { return hiddenSize; }
+
+ private:
+  /**
+   * VisiblePreAction function calculates the pre activation
+   * values given the hidden input units.
+   *
+   * @param input hidden unit neuron
+   * @param ouput visible unit pre-activation values
+   */
+  void VisiblePreActivation(InputDataType&& input, OutputDataType&& output);
+  /**
+   * HiddenPreActivation function calculates the pre activation
+   * values given the hidden input units.
+   *
+   * @param input visible unit neuron
+   * @param ouput hidden unit pre-activation values
+   */
+  void HiddenPreActivation(InputDataType&& input, OutputDataType&& output);
+
+ private:
+  //! Locally stored number of visible neurons
+  size_t visibleSize;
+  //! Locally stored number of hidden neurons
+  size_t hiddenSize;
+  //! Locally stored  Parameters of the network
+  InputDataType parameter;
+  //! Locally stored weight of the network
+  InputDataType weight;
+  //! Locally stored biases of the visible layer
+  OutputDataType visibleBias;
+  //! Locally stored biases of hidden layer
+  OutputDataType hiddenBias;
+  //! Locally-stored output of the preActivation function used in FreeEnergy
+  InputDataType preActivation;
+  //! Locally-stored corrupInput used for Pseudo-Likelihood
+  InputDataType corruptInput;
```

Comment:
Well this is a good question. There is actually dosen't make sense to have diffrent input and out datatypes. But the same logic applies for FFN layers as well.